cumulusci/tasks/bulkdata/snowfakery.py

```diff
@@ -631,133 +587,106 @@ def _cleanup_object_tables(self, engine, metadata):
-            + self.sets_being_generated
-            + self.sets_queued_for_loading
-            + self.sets_being_loaded
-        )
+class RunningTotals:
+    errors: int = 0
+    successes: int = 0
 
-    @property
-    def total_sets_working_on_or_uploaded(
-        self,
-    ):
-        return self.total_in_flight + self.sets_finished
-
-    def _display(self, detailed=False):
-        most_important_stats = [
-            "target_count",
-            "total_sets_working_on_or_uploaded",
-            "sets_finished",
-            "sets_failed",
-        ]
+    def __repr__(self):
+        return f"<{self.__class__.__name__} {self.__dict__}>"
 
-        queue_stats = [
-            "inprogress_generator_jobs",
-            "inprogress_loader_jobs",
-        ]
 
-        def display_stats(keys):
-            def format(a):
-                return f"{a.replace('_', ' ').title()}: {getattr(self, a):,}"
+class RulesFileAndRules(T.NamedTuple):
+    file: Path
+    rules: T.List[ChannelDeclaration]
 
-            return (
-                "\n"
-                + "\n".join(
-                    format(a)
-                    for a in keys
-                    if not a[0] == "_" and not callable(getattr(self, a))
-                )
-                + "\n"
-            )
 
-        rc = display_stats(most_important_stats)
-        rc += "\n   ** Workers **\n"
-        rc += display_stats(queue_stats)
-        if detailed:
-            rc += "\n   ** Internals **\n"
-            rc += display_stats(
-                set(dir(self)) - (set(most_important_stats) & set(queue_stats))
+def _read_channel_declarations_from_file(
+    loading_rules_file: Path,
+) -> RulesFileAndRules:
+    with loading_rules_file.open() as f:
+        decls = SObjectRuleDeclarationFile.parse_from_yaml(f)
+        channel_decls = decls.channel_declarations or []
+        assert isinstance(channel_decls, list)
+        return RulesFileAndRules(loading_rules_file, channel_decls)
+
+
+def read_channel_declarations(
+    recipe: Path, loading_rules_files: T.List[Path]
+) -> T.List[ChannelDeclaration]:
+    implicit_rules_file = infer_load_file_path(recipe)
+    if implicit_rules_file and implicit_rules_file.exists():
+        loading_rules_files = loading_rules_files + [implicit_rules_file]
+    # uniqify without losing order
+    loading_rules_files = list(dict.fromkeys(loading_rules_files))
+
+    rules_lists = [
+        _read_channel_declarations_from_file(file) for file in loading_rules_files
+    ]
+    rules_lists = [rules_list for rules_list in rules_lists if rules_list.rules]
+
+    if len(rules_lists) > 1:
+        files = ", ".join(
+            [f"{rules_list.file}: {rules_list.rules}" for rules_list in rules_lists]
+        )
+        msg = f"Multiple channel declarations: {files}"
+        raise TaskOptionsError(msg)
+    elif len(rules_lists) == 1:
+        return rules_lists[0].rules
+    else:
+        return []
+
+
+class ChannelConfig(T.NamedTuple):
```

Comment:
@prescod I probably won't remember what a channel is in 6 months. Could you add some docstrings to the top-level functions and classes added in this PR?